servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java

```diff
@@ -231,33 +232,41 @@ public boolean isAvailable() {
             return retainedAddresses.isEmpty() ? null : retainedAddresses.values();
         }
 
-        @Nullable
-        Collection<E> consume(final Collection<E> events) {
+        Collection<E> consumeAndFilter(final Collection<E> events) {
             if (retainedAddresses == NONE_RETAINED) {
+                List<E> toReturn = new ArrayList<>(events.size());
                 for (E e : events) {
                     if (e.isAvailable()) {
-                        activeAddresses.put(e.address(), e);
+                        if (activeAddresses.putIfAbsent(e.address(), e) == null) {
+                            // FIXME: what if it's a custom E that has more fields?
+                            toReturn.add(e);    // filter out duplicated addresses
+                        }
                     } else {
-                        activeAddresses.remove(e.address());
+                        if (activeAddresses.remove(e.address()) != null) {
+                            toReturn.add(e);   // filter out addresses that were already unavailable
+                        }
                     }
+                    // FIXME: remove addresses that are available and unavailable in the same events collection
                 }
-                return events;
+                return toReturn;
             }
 
             // we have seen an error, replace cache with new addresses and deactivate the ones which are not present
             // in the new list.
             for (E event : events) {
-                final R address = event.address();
                 if (event.isAvailable()) {
-                    activeAddresses.put(address, event);
+                    activeAddresses.put(event.address(), event);
                 } else {
-                    activeAddresses.remove(address);
+                    activeAddresses.remove(event.address());
                 }
             }
 
-            List<E> toReturn = new ArrayList<>(activeAddresses.values());
-            for (R address : activeAddresses.keySet()) {
-                retainedAddresses.remove(address);
+            // Worst case size is a sum of activeAddresses and retainedAddresses
+            List<E> toReturn = new ArrayList<>(activeAddresses.size() + retainedAddresses.size());
```

Comment:
I had something like the following in mind... this will only insert items which weren't previously available, and insert unavailables for items which were previously available (or are marked as available and unavailable in the current list).

```diff
diff --git a/servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java b/servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java
index c261fbe5d..a2d5c3247 100644
--- a/servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java
+++ b/servicetalk-http-api/src/main/java/io/servicetalk/http/api/DefaultServiceDiscoveryRetryStrategy.java
@@ -234,46 +234,34 @@ public final class DefaultServiceDiscoveryRetryStrategy<ResolvedAddress,
 
         Collection<E> consumeAndFilter(final Collection<E> events) {
             if (retainedAddresses == NONE_RETAINED) {
-                List<E> toReturn = new ArrayList<>(events.size());
                 for (E e : events) {
                     if (e.isAvailable()) {
-                        if (activeAddresses.putIfAbsent(e.address(), e) == null) {
-                            // FIXME: what if it's a custom E that has more fields?
-                            toReturn.add(e);    // filter out duplicated addresses
-                        }
+                        activeAddresses.put(e.address(), e);
                     } else {
-                        if (activeAddresses.remove(e.address()) != null) {
-                            toReturn.add(e);   // filter out addresses that were already unavailable
-                        }
+                        activeAddresses.remove(e.address());
                     }
-                    // FIXME: remove addresses that are available and unavailable in the same events collection
                 }
-                return toReturn;
+                return events;
             }
 
             // we have seen an error, replace cache with new addresses and deactivate the ones which are not present
             // in the new list.
+            List<E> toReturn = new ArrayList<>(activeAddresses.size() + retainedAddresses.size());
             for (E event : events) {
+                final R address = event.address();
                 if (event.isAvailable()) {
-                    activeAddresses.put(event.address(), event);
-                } else {
-                    activeAddresses.remove(event.address());
+                    if (activeAddresses.put(address, event) == null && retainedAddresses.remove(address) == null) {
+                        toReturn.add(event);
+                    }
+                } else if (activeAddresses.remove(address) != null || retainedAddresses.remove(address) != null) {
+                    toReturn.add(event);
                 }
             }
 
-            // Worst case size is a sum of activeAddresses and retainedAddresses
-            List<E> toReturn = new ArrayList<>(activeAddresses.size() + retainedAddresses.size());
-            for (Map.Entry<R, E> entry : activeAddresses.entrySet()) {
-                if (retainedAddresses.remove(entry.getKey()) == null) {
-                    toReturn.add(entry.getValue()); // Add only those new addresses that were not retained before
-                }
+            for (E event : retainedAddresses.values()) {
+                toReturn.add(flipAvailability.apply(event));
             }
 
-            if (!retainedAddresses.isEmpty()) {
-                for (E evt : retainedAddresses.values()) {
-                    toReturn.add(flipAvailability.apply(evt));
-                }
-            }
             retainedAddresses = noneRetained();
             return toReturn;
         }
```